src/metabase/api/routes.clj

```diff
@@ -61,50 +25,71 @@
-  (context "/segment"              [] (+auth segment/routes))
-  (context "/session"              [] session/routes)
-  (context "/setting"              [] (+auth setting/routes))
-  (context "/setup"                [] setup/routes)
-  (context "/slack"                [] (+auth slack/routes))
-  (context "/table"                [] (+auth table/routes))
-  (context "/task"                 [] (+auth task/routes))
-  (context "/testing"              [] (if (or (not config/is-prod?)
-                                              (config/config-bool :mb-enable-test-endpoints))
-                                        testing/routes
-                                        (fn [_ respond _] (respond nil))))
-  (context "/tiles"                [] (+auth tiles/routes))
-  (context "/transform"            [] (+auth transform/routes))
-  (context "/user"                 [] (+auth user/routes))
-  (context "/util"                 [] util/routes)
-  (route/not-found (constantly {:status 404, :body (deferred-tru "API endpoint does not exist.")})))
+(defn- pass-thru-handler
+  "A no-op handler that always passes thru to the next handler."
+  [_ respond _]
+  (respond nil))
+
+(defn- +enable-for-testing
+  "Wrap `handler` so it is only enabled for non-prod run configs, or if env var `MB_ENABLE_TEST_ENDPOINTS` is truthy."
+  [handler]
+  (if (or (not config/is-prod?)
+          (config/config-bool :mb-enable-test-endpoints))
+    handler
+    pass-thru-handler))
+
+(def ^:private lazy-ee-handler
+  "Lazy handler that attempts to load the EE-specific routes on the first request."
+  (lazy/handler
+   (fn []
+     (or (u/ignore-exceptions
+           (classloader/require 'metabase-enterprise.sandbox.api.routes)
+           (log/trace "Lazy-loaded Metabase Enterprise API routes")
+           (resolve 'metabase-enterprise.sandbox.api.routes/routes))
+         pass-thru-handler))))
+
+(def ^{:arglists '([request respond raise])} routes
+  "Ring routes for API endpoints."
+  (compojure/routes
+   lazy-ee-handler
+   (lazy/routes metabase.api
+    (+auth activity)
+    (+auth alert)
+    (+auth automagic-dashboards)
+    (+auth card)
+    (+auth collection)
+    (+auth dashboard)
+    (+auth database)
+    (+auth dataset)
+    (+auth email)
+    (+message-only-exceptions embed)
+    (+auth field)
+    geojson
+    (+auth ldap)
+    (+auth login-history)
+    (+auth metastore)
+    (+auth metric)
+    (+auth native-query-snippet)
+    (+apikey notify)
+    (+auth permissions)
+    (+auth preview-embed)
+    (+generic-exceptions public)
+    (+auth pulse)
+    (+auth revision)
+    (+auth search)
+    (+auth segment)
+    session
+    (+auth setting)
+    setup
+    (+auth slack)
+    (+auth table)
+    (+auth task)
+    (+enable-for-testing testing)
+    (+auth tiles)
+    (+auth transform)
+    (+auth user)
+    util)
+   (route/not-found (constantly {:status 404, :body (deferred-tru "API endpoint does not exist.")}))))
```

Comment:
I spent a long time thinking about how I'd like the code to look here. I think this is about as nice as it can get. I used this PR as an opportunity to make things a lot more concise than before and remove a bunch of the boilerplate. This macro has the added bonus of enforcing our pattern that API namespaces match the API routes e.g. `/api/card` endpoints live in `metabase.api.card` and `/api/dashboard` endpoints live in `metabase.api.dashboard`

Reply:
Actually it turns out `/api/preview_embed` was the one group of endpoints *not* following that pattern, so I went ahead and tweaked it so it follows the same pattern as everyone else. It's now `/api/preview-embed`